### PR TITLE
Dhl

### PR DESCRIPTION
**Pull request by aaronmwhitehead**

### CODE REVIEW

File: carriers/dhl.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+const request = require('request');
+const geography = require('../util/geography');
+const moment = require('moment-timezone');
+const checkDigit = require('../util/checkDigit');
+
+function DHL(options) {
+    this.isTrackingNumberValid = function(trackingNumber) {
+        // remove whitespace
+        trackingNumber = trackingNumber.replace(/\s/g, '');
+        trackingNumber = trackingNumber.toUpperCase();
```
Comment (javamatte) [inline carriers/dhl.js]: Looking at the valid tracking numbers in the test file, none have alpha characters. We probably don't need this line.

File: carriers/dhl.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            headers: {
+                'DHL-API-Key': `${options.DHL_API_Key}`
+            },
+            timeout: 5000
+        };
+
+        async.retry(function(callback) {
+            request(req, callback);
+        }, function(err, res) {
+            const response = JSON.parse(res.body);
```
Comment (javamatte) [inline carriers/dhl.js]: If you set `json: true` in the `req` options above, you won't need to explicitly parse out the body and it adds some headers that should be required (accepts: json etc).

File: carriers/dhl.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                //invalid credentials or invalid tracking number
+            } else if (response.status === 401 || response.status === 404) {
+                return callback(new Error (response.detail));
+            }
+
+            const results = {
+                events: []
+            };
+
+            const scanDetails = JSON.parse(res.body).shipments[0].events;
```
Comment (javamatte) [inline carriers/dhl.js]: This should probably be a `var` instead since we're modifying `scanDetails` next.

Comment (javamatte) [reply carriers/dhl.js]: Also, as above, if the `json: true` option is set, the JSON.parse() is already taken care of.

File: carriers/dhl.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    zip: scanDetail.location != undefined ? scanDetail.location.address.postalCode : ''
+                }
+                scanDetail.location = geography.addressToString(scanDetail.address);
+            });
+
+            // Get unqiue array of locations
+            const locations = Array.from(new Set(scanDetails.map(scanDetail => scanDetail.location)));
+
+            // Lookup each location
+            async.mapLimit(locations, 10, function (location, callback) {
```
Comment (javamatte) [inline carriers/dhl.js]: `function(` instead of `function (` throughout.
70, 71, 80, many in the test file

File: carriers/dhl.js (hunk 1 above)
Comment (javamatte) [reply carriers/dhl.js]: Whoops, looking at the tests we _are_ sending in some alpha characters in the invalidTrackingNumber list.
Carry on :)

File: test/carriers/dhl.js [hunk 5]
```diff
@@ -0,0 +1,357 @@
+const assert = require('assert');
```
Comment (javamatte) [inline test/carriers/dhl.js]: `function(` -> `function (` throughout

File: carriers/dhl.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                    const address = addresses.find(a => a && a.location === scanDetail.location);
+                    let timezone = 'America/New_York';
+
+                    if (address && address.timezone) {
+                        timezone = address.timezone;
+                    }
+
+                    const event = {
+                        address: scanDetail.address,
+                        date: new Date(moment.tz(`${scanDetail.timestamp}`, timezone).format('YYYY-MM-DDTHH:mm:ss')),
```
Comment (freshlogic) [inline carriers/dhl.js]: Shouldn't this be more like `date: moment.tz(scanDetail.timestamp, 'YYYY-MM-DDTHH:mm:ss', timezone).toDate(),`?

**Comment by freshlogic** [comment]
Needs updated docs in README

**Pull request merged**